postgres_db/src/download_queue.rs

```diff
@@ -103,11 +105,110 @@ fn enqueue_chunk(conn: &DbConnection, chunk: &[DownloadTask]) -> usize {
         panic!("Programming error: enqueue_chunk must be called with a chunk of size <= ENQUEUE_CHUNK_SIZE ({})", ENQUEUE_CHUNK_SIZE);
     }
 
+
+    // A1. Filter out URL x if it exists in downloaded_tarballs
+    let already_downloaded_urls: HashSet<_> = download_tarball::get_downloaded_urls_matching_tasks(conn, chunk).into_iter().collect();
+    let chunk: Vec<_> =  chunk.iter().filter(|t| !already_downloaded_urls.contains(&t.url)).collect();
+
+    // A2. If URL x didn't exist in downloaded_tarballs, enqueue into download_tasks
     diesel::insert_into(download_tasks)
         .values(chunk)
         .on_conflict_do_nothing()
         .execute(&conn.conn)
         .expect("Failed to enqueue downloads into DB")
+
+
+    // Note: we could do a transaction here, but instead if we consider
```

Comment:
How would we go about making the transaction? [This](https://www.postgresql.org/docs/current/tutorial-transactions.html) ?

Reply:
That's the core concept in SQL, and its available in Diesel with this: https://docs.diesel.rs/master/diesel/prelude/trait.Connection.html#method.transaction